### PR TITLE
Spec for the error format minitest expects from Module#undef_method

### PR DESCRIPTION
**Pull request by chrisseaton**
I put this as a Truffle spec as I wasn't sure how to describe or isolate the underlying rule. It's lifted straight from a failure in `minitest`, and fixes it, just not sure about the spec.

https://github.com/Shopify/truffleruby/issues/1

### CODE REVIEW

File: src/main/java/org/truffleruby/core/exception/CoreExceptions.java [hunk 1]
```diff
@@ -642,7 +642,7 @@ public DynamicObject nameErrorInstanceVariableNotDefined(String name, Object rec
     public DynamicObject nameErrorUndefinedMethod(String name, DynamicObject module, Node currentNode) {
         assert RubyGuards.isRubyModule(module);
         return nameError(
-                StringUtils.format("undefined method `%s' for %s", name, Layouts.MODULE.getFields(module).getName()),
+                StringUtils.format("undefined method `%s' for class `%s'", name, Layouts.MODULE.getFields(module).getName()),
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/exception/CoreExceptions.java]:
```
$ ruby -e 'class C; undef_method :foo; end'
-e:1:in `undef_method': undefined method `foo' for class `C' (NameError)

$ ruby -e 'module C; undef_method :foo; end'
-e:1:in `undef_method': undefined method `foo' for module `C' (NameError)
```

So it's `class` or `module` based on what it is, could you fix that?

File: spec/truffle/undef_method_spec.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+      def truffleruby_minitest_spec_stub
+        metaclass = class << self; self; end
+        metaclass.send :alias_method, :new_truffleruby_minitest_spec_stub_does_not_exist, :truffleruby_minitest_spec_stub_does_not_exist
+      ensure
+        metaclass.send :undef_method, :truffleruby_minitest_spec_stub_does_not_exist
+      end
+
+    end
+
+    -> { Time.truffleruby_minitest_spec_stub }.should raise_error(NameError, /undefined method `truffleruby_minitest_spec_stub_does_not_exist' for class `Time'/)
```
Comment (eregon) [inline spec/truffle/undef_method_spec.rb]: Could you add specs in `spec/ruby/core/module/undef_method_spec.rb` instead?
Specs under `spec/truffle` should only contain things that cannot run on MRI or where our behavior differs intentionally.

The spec for the message can much simpler, as shown by my examples in the other comment.
Line 59 of `spec/ruby/core/module/undef_method_spec.rb` already has an example which can be copy-pasted and used to assert the message.
Please also add one for a `Class.new`, as the message differs there.

If there is special behavior for singleton classes, then please add a spec for that too.

Comment (chrisseaton) [reply spec/truffle/undef_method_spec.rb]: > Could you add specs in spec/ruby/core/module/undef_method_spec.rb instead?
Specs under spec/truffle should only contain things that cannot run on MRI or where our behavior differs intentionally.

Yes, as I already said, the spec is only there as I wasn't sure how to write it best for `spec/ruby` and I was asking for advice.

**Comment by chrisseaton** [comment]
Done.

File: spec/ruby/core/class/undef_method_spec.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
+require_relative '../../spec_helper'
+
+describe "Class#undef_method" do
```
Comment (eregon) [inline spec/ruby/core/class/undef_method_spec.rb]: There is no method `undef_method` defined on Class, it's only defined on Module.
ruby/spec is not very strict about this (e.g., there are specs for SomeClass#dup even though it should be SomeClass#initialize_copy` or so), but I think in this case it's best to have them in the same file.
I'll move the spec.

File: src/main/java/org/truffleruby/core/module/ModuleFields.java [hunk 4]
```diff
@@ -447,9 +447,17 @@ public boolean removeMethod(String methodName) {
     public void undefMethod(RubyContext context, Node currentNode, String methodName) {
         final InternalMethod method = ModuleOperations.lookupMethodUncached(rubyModuleObject, methodName, null);
         if (method == null || method.isUndefined()) {
+            final DynamicObject moduleForError;
+
+            if (Layouts.CLASS.isClass(rubyModuleObject) && Layouts.CLASS.getIsSingleton(rubyModuleObject)) {
+                moduleForError = Layouts.CLASS.getAttached(rubyModuleObject);
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/module/ModuleFields.java]: We should have a spec for that special case, I'll add one.

**Comment by eregon** [comment]
I improved the specs and it revealed `Module#{to_s,inspect}` was broken for singleton classes, it returned an extra leading `#<Class:`.

**Comment by eregon** [comment]
Merged in 8c4783fd3931278556befa5529dcecc2748accc7

**Pull request merged**